### ISSUE TITLE
Different gradients in torch.matmul depending on input shape

### ISSUE THREAD

**Issue by dcasbol** [pytorch/pytorch#53263]
## 🐛 Bug

`torch.matmul` often returns different gradients for the same matrices, when the computation is done with an additional dimension (batched version). So if A and B are 2D matrices:

```python
C = torch.matmul(A,B)
D = torch.matmul(A.unsqueeze(0), B.unsqueeze(0)).squeeze(0)
```

Computing the gradient from C and D will give different results. This doesn't happens always, however, and it seems to be an issue related to numerical precision.

## To Reproduce

Make repeated executions of random matrix multiplications with both methods, but sharing the input values. The gradients will diverge at some point. I have prepared the following code for an accurate reproduction:

```python
import torch
import random

""" Do matrix-multiplications and gradient computation
several times. When gradients differ, compute total absolute
difference wrt. reference gradient computed in plain Python. """

def compute_accumulated_difference():
	err_matmul, err_bmm = 0.0, 0.0
	for _ in range(1000):
		N = random.randint(1,100)
		M = random.randint(1,100)
		L = random.randint(1,100)
		A = new_matrix(N,M)
		B = new_matrix(M,L)
		refs = matmul_grad(A,B)
		a = torch.tensor(A, dtype=torch.float)
		b = torch.tensor(B, dtype=torch.float)
		emm, ebmm = compare_methods(a, b, refs)
		err_matmul += emm
		err_bmm += ebmm
	print(err_matmul)
	print(err_bmm)

# Matrix multiplication functions in plain Python and native double format
def new_matrix(N,M):
	return [ [ random.gauss(0,1) for _ in range(M) ] for _ in range(N) ]

def transpose(A):
	return list(zip(*A))

def matmul(A,B):
	N, M, L = len(A), len(B), len(B[0])
	res = [ [ 0.0 for _ in range(L) ] for _ in range(N) ]
	for i in range(N):
		for j in range(L):
			res[i][j] = sum([ A[i][k]*B[k][j] for k in range(M) ])
	return res

def matmul_grad(A,B):
	N, L = len(A), len(B[0])
	G = [ [ 1.0 for _ in range(L) ] for _ in range(N) ]
	dA = matmul(G, transpose(B))
	dB = matmul(transpose(A), G)
	return dA, dB

# Functions in PyTorch to compare
def compute_matmul(a,b):
	return torch.matmul(a,b)

def compute_bmm(a,b):
	return torch.matmul(a.unsqueeze(0), b.unsqueeze(0)).squeeze(0)

def compute_grads(a,b,fn):
	a = a.detach().requires_grad_()
	b = b.detach().requires_grad_()
	c = fn(a,b)
	c.sum().backward()
	return a.grad, b.grad

def compare_methods(a,b,refs):
	res_1 = compute_grads(a,b,compute_matmul)
	res_2 = compute_grads(a,b,compute_bmm)
	err_1, err_2 = 0.0, 0.0
	for g1, g2, ref in zip(res_1, res_2, refs):
		if not (g1==g2).all():
			r = torch.tensor(ref, dtype=torch.double)
			err_1 += (g1.double()-r).abs().sum().item()
			err_2 += (g2.double()-r).abs().sum().item()
	return err_1, err_2

if __name__ == '__main__':
	compute_accumulated_difference()
```

## Expected behavior

Gradients shouldn't change their values because the shape is different, so computing `torch.matmul` or its batched version (unsqueezing inputs and squeezing outputs or `torch.bmm`) should give the exact same results.

## Environment

```
PyTorch version: 1.7.1+cpu
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2

Python version: 3.6 (64-bit runtime)
Is CUDA available: False
CUDA runtime version: No CUDA
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] adabelief-pytorch==0.2.1
[pip3] numpy==1.19.5
[pip3] pytorch-lightning==1.1.6
[pip3] torch==1.7.1+cpu
[pip3] torchfile==0.1.0
[conda] blas                      1.0                         mkl  
[conda] mkl                       2020.1                      217  
[conda] mkl-service               2.3.0            py38he904b0f_0  
[conda] mkl_fft                   1.1.0            py38h23d657b_0  
[conda] mkl_random                1.1.1            py38h0573a6f_0  
[conda] numpy                     1.18.5           py38ha1c710e_0  
[conda] numpy-base                1.18.5           py38hde5b4d6_0  
[conda] numpydoc                  1.1.0                      py_0
```

**Comment by vpirogov** [comment]
Looks like similar question was discussed in #51276.

**Comment by dcasbol** [comment]
Thanks, it looks like it could be related. In this case, forward results are exactly the same, but gradients are different. Errors in batched gradients add up to be an order of magnitude higher than without that extra dimension. Looking into the matmul implementation, it seems like the forward computation is recursive, so there should be no difference in the implementation. I don't know why this doesn't happen in the backward pass.

**Comment by zhangguanheng66** [comment]
cc @ngimel Do you have any ideas here?

**Comment by ngimel** [comment]
Unless you are seeing errors that are higher than would be expected for regular floating point accumulation errors, I don't think that this is a bug that should be fixed. Pytorch is calling into math libraries (mkl, cublas) for matmul computation, in particular, into batched matmul implementations or single matmul implementations in those libraries. Libraries themselves guarantee that the results will be the same. I don't think it's reasonable to make pytorch do another check on batch dimension and squeeze it if it's 1.

**Comment by dcasbol** [comment]
I am seeing errors that are an order of magnitude higher than I would have expected otherwise. This was discovered when a training started throwing NaNs after the change, so it is a reasonable concern. Anyways, something that I don't fully get is: if forward and backward operations are the same (only change matrices), how is it that forward always works as expected but there are divergences in the backward? Can't they just be merged to use the same implementation?

**Comment by ngimel** [comment]
All the matrix multiplications (forward and backward) are using the same implementation.

**Comment by dcasbol** [comment]
That is not what I have seen. Forward values match exactly but gradients do not. How could such a behaviour be explained otherwise? My impression is that the implementation used in the forward matrix-multiplication must be somehow different than the the implementation used in the backward. Am I missing something? As I understand it, the following code should either print out all `True` or all `False`, but not a mix.

```python
def fwd_and_grad(a, b, fn):
    a = a.detach().requires_grad_()
    b = b.detach().requires_grad_()
    c = fn(a,b)
    c.sum().backward()
    return c, a.grad, b.grad
    
a = torch.randn(100, 200)
b = torch.randn(200, 50)
res_1, a_grad_1, b_grad_1 = fwd_and_grad(a, b, lambda x, y: torch.matmul(x, y))
res_2, a_grad_2, b_grad_2 = fwd_and_grad(a, b, lambda x, y: torch.matmul(x.unsqueeze(0), y.unsqueeze(0)).squeeze(0))
print((res_1 == res_2).all()) # True
print((a_grad_1 == a_grad_2).all()) # False
print((b_grad_1 == b_grad_2).all()) # False
```

**Comment by dcasbol** [comment]
Hi all again. In my team we are facing significant difficulties due to this numeric instability, so I thought I could add some context to the issue to help support its importance.

We are trying to parallelize a meta-learning loop and it works quite well with 2D matrices. We can do up to hundreds of iterations, which means that we do hundreds of consecutive matrix multiplications. However, this is very troublesome with 3D matrices, leading to the method being unable to deal with more than 50 iterations. We initially spent more than 3 days investigating it and we traced the issue down to this numeric instability.

**Comment by ngimel** [comment]
Can't repro on master with mkl:
```
In [3]: def fwd_and_grad(a, b, fn):
   ...:     a = a.detach().requires_grad_()
   ...:     b = b.detach().requires_grad_()
   ...:     c = fn(a,b)
   ...:     c.sum().backward()
   ...:     return c, a.grad, b.grad
   ...: 
   ...: a = torch.randn(100, 200)
   ...: b = torch.randn(200, 50)
   ...: res_1, a_grad_1, b_grad_1 = fwd_and_grad(a, b, lambda x, y: torch.matmul(x, y))
   ...: res_2, a_grad_2, b_grad_2 = fwd_and_grad(a, b, lambda x, y: torch.matmul(x.unsqueeze(0), y.unsqueeze(0)).squeeze(0))
   ...: print((res_1 == res_2).all()) # True
   ...: print((a_grad_1 == a_grad_2).all()) # False
   ...: print((b_grad_1 == b_grad_2).all()) # False
tensor(True)
tensor(True)
tensor(True)
```
```
In [7]: print(*torch.__config__.show().split("\n"), sep="\n")
PyTorch built with:
  - GCC 8.4
  - C++ Version: 201402
  - Intel(R) Math Kernel Library Version 2020.0.2 Product Build 20200624 for Intel(R) 64 architecture applications
  - Intel(R) MKL-DNN v1.8.0 (Git Hash 2c8d20640d5068e2d85e378b266644fe86220e84)
  - OpenMP 201511 (a.k.a. OpenMP 4.5)
  - NNPACK is enabled
  - CPU capability usage: AVX2
  - CUDA Runtime 11.0
  - NVCC architecture flags: -gencode;arch=compute_70,code=sm_70
  - CuDNN 8.0.4
  - Magma 2.5.2
```
